[18.0][MIG] analytic_hr_department_restriction: Migration to 18.0

victoralmau opened the pull request:
Migration to 18.0

Please @pedrobaeza and @carlos-lopez-tecnativa can you review it?

@Tecnativa TT56200

victoralmau committed:
[ADD] analytic_hr_department_restriction: New module

TT55725

[UPD] Update analytic_hr_department_restriction.pot

[BOT] post-merge updates

victoralmau committed:
[IMP] analytic_hr_department_restriction: pre-commit auto fixes

victoralmau committed:
[MIG] analytic_hr_department_restriction: Migration to 18.0

TT56200

carlos-lopez-tecnativa reviewed (commented):
If the user does not have the `Analytic Accounting (without department)` group, I get an error.
I remember that this error did not occur in version 16.

![image](https://github.com/user-attachments/assets/7b7a869e-eb24-461d-a97a-e7a0ba4e477e)

Plan without department
![image](https://github.com/user-attachments/assets/b01021ab-e131-4347-8e4f-9888d4b73ec1)

![image](https://github.com/user-attachments/assets/aa44206e-36a9-469b-b768-fdb2d8ede737)

pedrobaeza commented:
/ocabot migration analytic_hr_department_restriction

victoralmau commented:
> If the user does not have the `Analytic Accounting (without department)` group, I get an error. I remember that this error did not occur in version 16.

Fixed.

carlos-lopez-tecnativa reviewed (approved):
LGTM

pedrobaeza commented on analytic_hr_department_restriction/models/account_analytic_plan.py:
```diff
@@ -2,13 +2,23 @@
-        string="Department",
-        domain="['|',('company_id', '=?', company_id),('company_id', '=', False)]",
-    )
+    department_id = fields.Many2one(comodel_name="hr.department", string="Department")
```
Why removing the domain?

victoralmau replied:
The company field no longer exists in the plans since v17.

pedrobaeza replied:
Yeah, but this is `hr.department`, which has company, so this should be converted to company dependent or be a many2many.

victoralmau replied:
Ok, change to many2many.

pedrobaeza commented on analytic_hr_department_restriction/models/ir_rule.py:
```diff
@@ -0,0 +1,59 @@
+            )
+            if user.has_group(group2):
+                extra_domain = [
+                    (department_f_name, "child_of", user.employee_ids.department_id.ids)
```
`child_of` works properly with many2many?

victoralmau replied:
It doesn't work correctly, I'm going to follow the way of making the company_dependent field https://github.com/OCA/account-analytic/pull/778#discussion_r2076890490

pedrobaeza commented on analytic_hr_department_restriction/migrations/18.0.1.0.0/post-migration.py:
```diff
@@ -0,0 +1,43 @@
+@openupgrade.migrate()
+def migrate(env, version):
+    """Set the department_id value in the project company."""
+    if not column_exists(env.cr, "account_analytic_plan", "old_department_id"):
```
Why not using openupgradelib method for doing the company dependent transformation?

victoralmau replied:
What would be the method to do this automatically? At https://github.com/OCA/helpdesk/pull/611 it was necessary to do this whole process to get the values right.

pedrobaeza replied:
https://github.com/OCA/openupgradelib/blob/20a96fff6490c379598926ce2d737bb83b3b6af6/openupgradelib/openupgrade.py#L3533

victoralmau committed:
[IMP] analytic_hr_department_restriction: Multi-company compatibility

Example of use case:
- User with access to several companies.
- Employee defined in one company
- Department and plan with no company defined
- Purchase order linked to company A (different from the user's employee).
- No access error should be displayed and the user should be able to see
the corresponding plan/analytic account even if the employee's company
is not selected.

TT56276

luisDIXMIT commented on analytic_hr_department_restriction/README.rst:
```diff
@@ -0,0 +1,115 @@
+Users will only be able to view plans and analytical accounts linked to
+their employee's department or those with no department defined.
+
+Users with the "Invoicing / Billing" permission will be able to view all
```
Please, can you update it to Invoicing / Invoicing? group_account_invoice is now named "Invoicing" https://github.com/odoo/odoo/blob/752ba2927e0b61cd6315e5c780a5292a6e99fe17/addons/account/security/account_security.xml#L60

Same in lines 59 and 64

luisDIXMIT reviewed (commented):
Hi @victoralmau , I have a question. A user without the Invoicing / Invoicing permission should only be able to access those analytic accounts/plans that belong to their department. However, a user without this permission can't access anything related to invoicing/accounting, let alone analytic accounting. What am I doing wrong? Thanks.

victoralmau commented:
> Hi @victoralmau , I have a question. A user without the Invoicing / Invoicing permission should only be able to access those analytic accounts/plans that belong to their department. However, a user without this permission can't access anything related to invoicing/accounting, let alone analytic accounting. What am I doing wrong? Thanks.

This module only limits the analytical plans/accounts that the user can access.

It is possible to list the analytical plans/accounts from different sides, for example, if you have the purchase permission and the “Analytical accounting” permission you can set an analytical distribution on the purchase order lines, and there you are showing the “limited” analytical plans/accounts.

luisDIXMIT reviewed (approved):
Thanks you for the explanation @victoralmau I tested it on runboat and LGTM!

pedrobaeza reviewed (approved):
/ocabot merge nobump
